src/ast.h

```diff
@@ -0,0 +1,299 @@
+/*
+ * (ast.h | 8 Mar 19 | Amr Anwar)
+ * 
+ * The abstract syntax trees structs
+ * 
+*/
+
+#ifndef ast_h
+#define ast_h
+
+#include <stdint.h>
+
+#include "list.h"
+#include "lexer.h"
+
+typedef struct expr expr;
+typedef struct pattern pattern;
+
+typedef struct {
+    list *stmts;
+} piece;
+
+typedef enum {
+    let_stmt_type,
+    fn_stmt_type,
+    ret_stmt_type,
+    expr_stmt_type,
+    fixed_stmt_type
+} stmt_type;
+
+typedef enum {
+    BREAK,
+    CONTINUE,
+} fixed_stmt;
+
+typedef struct {
+    char* ident;
+    list* patts;
+} param_list;
+
+typedef struct {
+    char *n_ident;
+    list *param_list;
+    piece *fn_p;
+} fn_stmt;
+
+typedef struct {
+    char *ident;
+    expr *expression;
+} let_stmt;
+
+typedef struct {
+    expr *expression;
+} ret_stmt;
+
+typedef struct {
+    expr *expression;
+} expr_stmt;
+
+typedef struct {
+    stmt_type type;
+    union {
+        let_stmt * ls;
+        fn_stmt *fns;
+        ret_stmt *rs;
+        expr_stmt *es;
+        fixed_stmt fs;
+    } obj;
+} stmt;
+
+typedef enum {
+    lit_expr_type,
+    prefix_expr_type,
+    infix_expr_type,
+    index_expr_type,
+    access_expr_type,
+    group_expr_type,
+    call_expr_type,
+    if_expr_type,
+    for_expr_type,
+    while_expr_type,
+    match_expr_type,
+    ident_expr_type,
+} expr_type;
+
+typedef struct {
+    expr *expression;
+    char *n_ident;
+} access_expr;
+
+typedef struct {
+    token_type op;
+    expr *left;
+    expr *right;
+} infix_expr;
+
+typedef struct {
+    token_type op;
+    expr *expression;
+} prefix_expr;
+
+typedef struct {
+    expr *outer_e;
+    expr *inner_e;
+} index_expr;
+
+typedef struct {
+    expr *group_e;
+} group_expr;
+
+typedef struct {
+    expr *expression;
+    list *exprs;
+} expr_list;
+
+typedef struct {
+    expr *expr;
+    list *expression_l;
+} call_expr;
+
+typedef struct {
+    expr *if_e;
+    piece *if_p;
+    expr *elif_e; 
+    piece *elif_p;
+    piece *else_p;
+} if_expr;
```

Comment:
this is wrong. `if` expression could have multiple `elif` branches.
you could define a new `elif` struct as
```c
typedef struct {
    expr *cond;
    piece *body;
} elif_brach;

typdef struct {
    expr *cond;
    piece *body;
    list *elifs /* list of elif branches */
    piece *alter;
} if_expr;
```
or use list of `expr` and list of `piece` inside the `if_expr` struct directly
```c
typdef struct {
    expr *cond;
    piece *body;
    list *conds;   /* list of expr. NULL if there is no elif */
    list  *bodies;  /* list of piece. NULL if there is no elif */
    piece *alter;
} if_expr;
```
it's your call.
note the new naming.